pkg/job_controller/job.go

```diff
@@ -324,6 +332,116 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
 	return result, nil
 }
 
+func (jc *JobController) ReconcileCron(meta metav1.Object, runtimeObj runtime.Object, policy *apiv1.RunPolicy) error {
+	cronJob := &appv1.Cron{}
+	cronEnabled := policy != nil && policy.CronPolicy != nil
+	if err := jc.Client.Get(context.Background(), types.NamespacedName{
+		Namespace: meta.GetNamespace(),
+		Name:      meta.GetName(),
+	}, cronJob); err != nil {
+		if errors.IsNotFound(err) {
+			if !cronEnabled {
+				//Do not need to run regularly
+				return nil
+			}
+			instance, err := jc.createCronInstance(policy, runtimeObj)
+			if err != nil {
+				return err
+			}
+			return jc.Client.Create(context.Background(), instance)
+		}
+		return err
+	}
+	if !cronEnabled {
+		// Job no longer needs to run regularly, delete the originally created cronJob
+		return jc.Client.Delete(context.Background(), cronJob)
+	}
+	// If cronPolicy or cronTemplate change, the cronJob needs to be updated
+	newCronJob := cronJob.DeepCopy()
+	newInstance, err := jc.createCronInstance(policy, runtimeObj)
+	if err != nil {
+		return err
+	}
+	if !reflect.DeepEqual(cronJob.Spec, newInstance.Spec) {
+		newCronJob.Spec.CronTemplate = newInstance.Spec.CronTemplate
+		return jc.Client.Patch(context.Background(), newCronJob, client.MergeFrom(cronJob))
+	}
+	return nil
+}
+
+func (jc *JobController) createCronInstance(policy *apiv1.RunPolicy, runtimeObj runtime.Object) (*appv1.Cron, error) {
+	tempPolicy := policy.DeepCopy()
+	defer func() {
+		policy.CronPolicy = tempPolicy.CronPolicy
```

Comment:
why is this needed ? already made a deepcopy

Reply:
Encode newMeta because the cron expression in newMeta has been emptied, so the job template in the newly created cron job has no cron expression. Otherwise the next reconcile job will create a new cron job with cron expressions, resulting in the creation of loops.
RightWay: PytorchJob(cron) --> CronJob(pytorchJob without cron exp) --> PytorchJob
WrongWay: PytorchJob(cron) --> CronJob(pytorchJob with cron exp)  -->  PytorchJob(cron) --> cronJob(pytorchJob with cron exp) --> loop

Reply:
my question is why this defer func is needed